File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/queue/P355_DesignTwitter.py
import heapq
from typing import List


class Twitter:
    timestamp = 1

    def __init__(self):
        self.user_map = {}

    def postTweet(self, userId: int, tweetId: int) -> None:
        if userId not in self.user_map:
            self.user_map[userId] = User(userId)
        self.user_map[userId].postTweet(tweetId)

    def getNewsFeed(self, userId: int) -> List[int]:
        if userId not in self.user_map:
            return []
        user = self.user_map[userId]
        heap = []
        for followeeId in user.followed:
            if followeeId in self.user_map:
                follow_user = self.user_map[followeeId]
                if follow_user.head is not None:
                    heapq.heappush(heap, (-follow_user.head.timestamp, follow_user.head))
        res = []
        while heap and len(res) < 10:
            _, tweet = heapq.heappop(heap)
            res.append(tweet.tweetId)
            if tweet.next is not None:
                heapq.heappush(heap, (-tweet.next.timestamp, tweet.next))
        return res

    def follow(self, followerId: int, followeeId: int) -> None:
        if followerId not in self.user_map:
            self.user_map[followerId] = User(followerId)
        if followeeId not in self.user_map:
            self.user_map[followeeId] = User(followeeId)
        self.user_map[followerId].follow(followeeId)

    def unfollow(self, followerId: int, followeeId: int) -> None:
        if followerId in self.user_map:
            self.user_map[followerId].unfollow(followeeId)


class User:
    def __init__(self, userId: int):
        self.userId = userId
        self.followed = set()
        self.head = None
        self.follow(userId)

    def follow(self, followeeId: int):
        self.followed.add(followeeId)

    def unfollow(self, followeeId: int):
        if self.userId != followeeId:
            self.followed.discard(followeeId)

    def postTweet(self, tweetId: int, ):
        tweet = Tweet(tweetId, Twitter.timestamp)
        Twitter.timestamp += 1
        tweet.next = self.head
        self.head = tweet


class Tweet:
    def __init__(self, tweetId: int, timestamp: int):
        self.tweetId = tweetId
        self.timestamp = timestamp
        self.next = None

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/queue/P355_DesignTwitter.java
package com.dsa.leetcode.labuladong.queue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

public class P355_DesignTwitter {
    class Twitter {
        int timestamp;
        Map<Integer, User> userMap;

        public Twitter() {
            timestamp = 1;
            userMap = new HashMap<>();
        }

        public void postTweet(int userId, int tweetId) {
            if (!userMap.containsKey(userId)) {
                userMap.put(userId, new User(userId));
            }
            userMap.get(userId).postTweet(tweetId);
        }

        public List<Integer> getNewsFeed(int userId) {
            if (!userMap.containsKey(userId)) {
                return new ArrayList<>();
            }
            User user = userMap.get(userId);
            PriorityQueue<Tweet> heap = new PriorityQueue<>((a, b) -> b.timestamp - a.timestamp);
            for (Integer id : user.followed) {
                User followee = userMap.get(id);
                if (followee.head != null) {
                    heap.add(followee.head);
                }
            }

            List<Integer> res = new LinkedList<>();
            while (!heap.isEmpty() && res.size() < 10) {
                Tweet poll = heap.poll();
                res.add(poll.tweetId);
                if (poll.next != null) {
                    heap.add(poll.next);
                }
            }
            return res;
        }

        public void follow(int followerId, int followeeId) {
            if (!userMap.containsKey(followerId)) {
                userMap.put(followerId, new User(followerId));
            }
            if (!userMap.containsKey(followeeId)) {
                userMap.put(followeeId, new User(followeeId));
            }
            userMap.get(followerId).follow(followeeId);
        }

        public void unfollow(int followerId, int followeeId) {
            if (userMap.containsKey(followerId)) {
                userMap.get(followerId).unfollow(followeeId);
            }
        }

        // user
        class User {
            int userId;
            // store followed
            Set<Integer> followed;
            Tweet head;

            public User(int userId) {
                this.userId = userId;
                followed = new HashSet<>();
                head = null;
                // must follow oneself
                follow(userId);
            }

            public void postTweet(int tweetId) {
                Tweet tweet = new Tweet(tweetId, timestamp);
                timestamp++;
                tweet.next = head;
                head = tweet;
            }

            public void follow(int followeeId) {
                followed.add(followeeId);
            }

            public void unfollow(int followeeId) {
                // can't unfollow oneself
                if (userId != followeeId) {
                    followed.remove(followeeId);
                }
            }

        }

        // // Tweet
        class Tweet {
            int tweetId;
            int timestamp;
            Tweet next;

            public Tweet(int tweetId, int timestamp) {
                this.tweetId = tweetId;
                this.timestamp = timestamp;
            }
        }
    }
}
